R/mod-file-summary.R

```diff
@@ -0,0 +1,326 @@
+        visualize_data_types(file_data()[[input$file_to_summarize]])
+      })
+      file_summary <- data_summary(file_data()[[input$file_to_summarize]])
+      column_defs <- get_column_definitions(file_summary)
+      output$data_details <- reactable::renderReactable({
+        reactable::reactable(
+          file_summary,
+          highlight = TRUE,
+          searchable = TRUE,
+          resizable = TRUE,
+          columns = column_defs
+        )
+      })
+    }
+  })
+}
+
+#' @title Visualize data types
+#'
+#' @description Visualize the data class types, including
+#' missing data, using the visdat and ggplot
+#' packages.
+#'
+#' @noRd
+#' @param data Dataframe or tibble with file data.
+#' @return ggplot from [visdata::vis_dat()]
+visualize_data_types <- function(data) {
+  if (!inherits(data, "tbl_df") && !inherits(data, "data.frame")) {
+    return(NULL)
+  }
+  visdat::vis_dat(data) +
+    ggplot2::theme(text = ggplot2::element_text(size = 16))
+}
+
+#' @title Get summary of data
+#'
+#' @description Get a summary of the data using the skimr
+#' package, along with a custom function that appends an
+#' extra column with a string showing each value for
+#' a given variable and the number of times the value
+#' appears. Returns a tibble with the data.
+#'
+#' @noRd
+#' @param data Dataframe or tibble with file data.
+#' @return Tibble with summary information.
+data_summary <- function(data) {
+  if (!inherits(data, "tbl_df") && !inherits(data, "data.frame")
+  || nrow(data) == 0) {
+    return(NULL)
+  }
+  data_sum <- tibble::as_tibble(skimr::skim(data))
+
+  # Cut out excess info from skimr results
+  # Different classes result in different columns; get only ones that exist
+  desired_cols <- c(
+    "skim_variable",
+    "skim_type",
+    "n_missing",
+    "complete_rate",
+    "character.n_unique",
+    "numeric.mean",
+    "numeric.sd",
+    "numeric.hist"
+  )
+  data_sum <- data_sum[, names(data_sum) %in% desired_cols]
+  data_sum <- tibble::add_column(data_sum, value_occurrence = NA)
+  for (var in data_sum$skim_variable) {
+    var_col <- which(names(data) == var)
+    data_sum$value_occurrence[data_sum$skim_variable == var] <-
+      summarize_values(data[[var_col]])
+  }
+  data_sum
+}
+
+#' @title Summarize values present
+#'
+#' @description Get a list of values present and
+#' the number of times each variable appeared.
+#'
+#' @noRd
+#' @param values The values to summarize in a list.
+#' @return String with the form "value1 (2), value2 (4)",
+#'   where the value is given with the number of
+#'   occurrences in parenthesis.
+summarize_values <- function(values) {
+  if (is.null(values)) {
+    return(NULL)
+  }
+  val_sum <- list()
+  for (value in unique(values)) {
+    x_appeared <- length(which(values == value))
+    val_sum <- append(
+      val_sum,
+      glue::glue("{value} ({x_appeared})")
+    )
+  }
+  val_sum_string <- glue::glue_collapse(val_sum, sep = ",  ")
+  val_sum_string
```

Comment:
I had not used it before. I went ahead and added it, especially since it clears up an annoyance with creating a variable specifically to use in a glue statement (R thinks the variable isn't being used anywhere). Thank you for the suggestion!